[quant][graph] PyTorch Graph Mode Quantization API

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26390 [quant][graph] PyTorch Graph Mode Quantization API**

Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D17645132](https://our.internmc.facebook.com/intern/diff/D17645132)

jerryzh168 committed:
[quant][graph] PyTorch Graph Mode Quantization API

Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/quantization/quantize_script.py:
```diff
@@ -0,0 +1,33 @@
+    torch._C._jit_pass_fold_convbn(model._c)
+    prepare_script(model, qconfig_dict, True)
+    print(model._c._get_method('forward').graph)
+    print(model._c._get_module('fc1')._get_method('forward').graph)
```
Debug leftovers?

ZolotukhinM commented on torch/quantization/quantize_script.py:
```diff
@@ -0,0 +1,33 @@
+
+def quantize_script(model, qconfig_dict, run_fn, run_args, inplace=False):
+    if not inplace:
+        model = clone_script(model)
```
I think we have `module.copy()` in python - we should probably use that here instead.

ZolotukhinM commented on torch/quantization/quantize_script.py:
```diff
@@ -0,0 +1,33 @@
+    prepare_script(model, qconfig_dict, True)
+    print(model._c._get_method('forward').graph)
+    print(model._c._get_module('fc1')._get_method('forward').graph)
+    run_fn(model._c._get_method('forward'), *run_args)
```
We probably need to check that the input module has the 'forward' method.

ZolotukhinM reviewed (commented):
Looks good to me overall, and I'm very happy to see that landing! I'll let others to review it as it's a user-facing API, but I left a couple of comments inline.

fmassa commented on test/test_quantization.py:
```diff
@@ -610,6 +615,37 @@ def checkQuantized(model):
 
 
+@unittest.skipIf(
+    not torch.fbgemm_is_cpu_supported(),
```
Nit: I would expect `fbgemm_is_cpu_supported` to be private, or live in a different namespace as `torch`, but I suppose this is an artifact on how we bind operators in Python?

jerryzh168 replied:
this is a native function:  https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L1429
why should it be private?

dzhulgakov replied:
it's an implementation detail - polluting the global namespace is bad. We should just bind it directly to _C with pybind and expose through torch.quantization.backends.bla (or just reuse the available engines you have in another diff).

At the very least, we should prepend an underscore to it. But it all deserves a separate diff probably.

dzhulgakov replied:
Btw, I fixed it in a separate PR: #26840

dzhulgakov commented on torch/quantization/quantize_script.py:
```diff
@@ -0,0 +1,33 @@
+    torch._C._jit_pass_insert_quant_dequant(model._c, 'forward', True)
+    return model
+
+def quantize_script(model, qconfig_dict, run_fn, run_args, inplace=False):
```
naming random thought - should it be `torch.jit.quantize`? :) cc @gchanan @jamesr66a

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -610,6 +615,37 @@ def checkQuantized(model):
+
+        qconfig_dict = {
+            '': QConfig(
+                activation=torch.jit.script(default_observer()())._c,
```
no `_c` please. And you can put this transformation inside the quantize() API instead of asking users to do it (including invoking `torch.jit.script`).

Also, I'd like to highlight it again - we shouldn't rely on quantizers being scriptable. There's nothing really preventing us calling back into python implementation (perf is not critical in these cases I suppose). It's really convenient for the users and would make eager and graph apis even closer.

jerryzh168 replied:
> Also, I'd like to highlight it again - we shouldn't rely on quantizers being scriptable. There's nothing really preventing us calling back into python implementation (perf is not critical in these cases I suppose). It's really convenient for the users and would make eager and graph apis even closer.

do you mean observers?

jerryzh168 replied:
supporting non-scriptable observers will be something we plan to do after lockdown

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
more comments everyone?

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov commented on torch/quantization/__init__.py:
```diff
@@ -15,13 +16,17 @@ def default_eval_fn(model, calib_data):
     'propagate_qconfig', 'add_quant_dequant', 'add_observer', 'swap_module',
     'default_eval_fn',
+    # Top level API for graph mode quantization
+    'quantize_script',
```
since graph mode is not yet ready for public light yet - can we remove it from __all__ here and maybe add underscore to `quantize_script` (this might be optional). We can codemod it back for 1.3

dzhulgakov commented on torch/quantization/quantize_script.py:
```diff
@@ -0,0 +1,36 @@
+import torch
+from .QConfig import QConfig
+
+def prepare_script(model, qconfig_dict, inplace=False):
```
is there a way to check that argument is the script module?

jerryzh168 replied:
I'll use type to check it

dzhulgakov commented on torch/quantization/quantize_script.py:
```diff
@@ -0,0 +1,36 @@
+    torch._C._jit_pass_insert_quant_dequant(model._c, 'forward', True)
+    return model
+
+def script_qconfig(qconfig):
```
add TODO please that we want also non-scriptable observers support

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov commented on torch/quantization/quantize_script.py:
```diff
@@ -0,0 +1,44 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
you could have probably called the whole file `_quantize_script` then you don't need to rename individual functions.

dzhulgakov reviewed (approved):
Looks good

jerryzh168 committed:
Update on "[quant][graph] PyTorch Graph Mode Quantization API"


Summary:
`quantize_script`: top level API for graph mode quantization

Test Plan:
there are some known issues, we can enable test after all known issues are fixed.

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
